### DISCUSSION TITLE
p/18d81a08c6aa4908aa22979e44b14660/

### DISCUSSION THREAD

**Discussion by giscus** [rxliuli/blog#18] [Announcements]
# p/18d81a08c6aa4908aa22979e44b14660/

这周发生了许多事情，但最重要的还是 chatgpt api 的上线。这次上线引发的社区疯狂的各种应用简直前所未有，吾辈这两天甚至开始焦虑，感觉正在发生一件开天辟地的大事，但自己却未能正确的参与其中。有人说吾辈是典型的 fomo 症状，或许正是如此，但也正是因为亲身体验了 chatgpt plus，才会相信它真的有这么好用，所有涉及到文本的事情都会受到影响。有人把它比喻为 iphone4 的发布，吾

https://blog.rxliuli.com/p/18d81a08c6aa4908aa22979e44b14660/

**Comment by miaopass-future** [comment]
chatGPT是真的好用啊，辅助解决代码问题，辅助查找解决办法。不过价格还是挺贵的

**Reply by rxliuli** [reply]
是的，但主要还是 ip 限制有点烦，国内不能用而且也不能使用香港的（最快的代理）而需要使用美国的（相对较慢），所以吾辈自己开始基于 chatgpt api 自己实现客户端了

**Reply by miaopass-future** [reply]
已经有api了么，之前的openai有点呆，特别容易问呆，不好用，没chat的智能

**Reply by rxliuli** [reply]
是的，chatgpt 3.5 的模型通过 api 开放了，不过实际上与网页上的仍然有些差异，很奇怪。

**Reply by rxliuli** [reply]
你可以试试 https://chat.ai-assist.moe/，与 chatgpt 网页版做对比

**Reply by miaopass-future** [reply]
今天看到你的这篇文章已经用了呢，Ai带我翻了下axios的底层实现，但是貌似问呆了，死活问不出axios的 async的实现。

**Reply by miaopass-future** [reply]
这个是基于chatgpt 3.5的api实现的么

**Reply by rxliuli** [reply]
chatgpt 官方网页版可以做到么？

**Reply by rxliuli** [reply]
> 这个是基于 chatgpt 3.5 的 api 实现的么

对，源码在 https://github.com/rxliuli/ai-assist

**Reply by miaopass-future** [reply]
也没问出来，感觉在胡说八道

**Reply by miaopass-future** [reply]
非常抱歉，我之前的回答仍然存在错误。在 Axios 的 0.24.0 版本中，async 选项的实现实际上是通过 XMLHttpRequest 对象的 open 方法来实现的。具体来说，如果 config.async 为 false，则在调用 open 方法时将其设置为 false，以使用同步请求模式。以下是实现该功能的代码示例：

javascript
Copy code
// Send the request
xhr.open(
  config.method.toUpperCase(),
  buildURL(config.url, config.params, config.paramsSerializer),
  config.async !== false, // Set the async option
  username,
  password
);

// ...

// Send the request
xhr.send(config.data);
如上所示，在使用 XMLHttpRequest 对象发送请求时，可以将 config.async 属性的值作为 open 方法的第三个参数传递，从而实现同步或异步请求模式。如果 config.async 为 false，则该参数的值将被设置为 false，以使用同步请求模式。否则，该参数的值将被设置为默认值 true，以使用异步请求模式。

**Reply by miaopass-future** [reply]
在 Axios 的 0.24.0 版本中，async 选项的实现位于 lib/adapters/xhr.js 文件中的 send 函数内，具体代码如下所示：

javascript
Copy code
if (config.async === false) {
  // Send the request synchronously
  send();
} else {
  // Send the request asynchronously
  xhr.timeout = config.timeout;

  xhr.onreadystatechange = function handleLoad() {
    if (!xhr || xhr.readyState !== 4) {
      return;
    }

    // ...

    if (requestError) {
      reject(createError(
        'Network Error',
        config,
        null,
        xhr
      ));
    } else {
      resolve({
        data: responseData,
        status: xhr.status,
        statusText: xhr.statusText,
        headers: responseHeaders,
        config,
        request: xhr
      });
    }
  };

  // ...

  // Send the request
  xhr.send(config.data);
}
如果 config.async 为 false，则使用 XMLHttpRequest 的同步模式发送请求，否则使用默认的异步模式发送请求。在异步模式下，XHR 对象的 onreadystatechange 事件将被设置为一个回调函数，该函数在请求状态发生变化时被调用，以检查请求是否完成。如果请求成功，resolve 函数将被调用，并返回包含响应数据、状态码、响应头和配置等信息的 JavaScript 对象。如果请求失败，则将使用 createError 函数创建一个错误对象，并使用 reject 函数将其传递给 Promise 链的下一个 catch 方法。

**Reply by miaopass-future** [reply]
这些是AI恢复的，但是没有找到相关代码的位置，还是吃的代码不够多，hha

**Reply by rxliuli** [reply]
嗯~ o(*￣▽￣*)o
